### PR TITLE
Add python IM event client code

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Add python IM event client support
Todo: Add python API to read attribute and event at the same time.
#### Change overview
See above
#### Testing
Compilation
Add E2E python tests since IM server integration( PR#12059), and currently we cannot retrieve events.

### CODE REVIEW

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 1]
```diff
@@ -471,6 +471,72 @@ async def ReadAttribute(self, nodeid: int, attributes: typing.List[typing.Union[
...
+        typing.Tuple[int, typing.Type[ClusterObjects.ClusterEventDescriptor]]
+    ]]):
+        '''
+        Read a list of events from a target node
+
+        nodeId: Target's Node ID
+        events: A list of tuples of varying types depending on the type of read being requested:
+            ():                                         Endpoint = *,           Cluster = *,          Event = *
+            (Clusters.ClusterA):                        Endpoint = *,           Cluster = specific,   Event = *
+            (Clusters.ClusterA.EventA):             Endpoint = *,           Cluster = specific,   Event = specific
```
Comment (erjiaqing) [inline src/controller/python/chip/ChipDeviceCtrl.py:493]: Nit: make these lines aligned

**Review by erjiaqing — COMMENTED** [review]
Should be OK as long as server side change are not ready, need to add some TODO to indicate this state.

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 1 above)
Comment (yunhanw-google) [reply src/controller/python/chip/ChipDeviceCtrl.py]: Done, thanks

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 2]
```diff
@@ -510,7 +584,19 @@ def ZCLReadAttribute(self, cluster, attribute, nodeid, endpoint, groupid, blocki
         result = asyncio.run(self.ReadAttribute(nodeid, [(endpoint, req)]))
         path = ClusterAttribute.AttributePath(
             EndpointId=endpoint, Attribute=req)
-        return im.AttributeReadResult(path=im.AttributePath(nodeId=nodeid, endpointId=path.EndpointId, clusterId=path.ClusterId, attributeId=path.AttributeId), status=0, value=result[path].Data.value)
+        return im.AttributeReadResult(path=im.AttributePath(nodeId=nodeid, endpointId=path.EndpointId, clusterId=path.ClusterId, attributeId=path.AttributeId), status=0, value=result['Attributes'][path].Data.value)
+
+    def ZCLReadEvent(self, cluster, event, nodeid, endpoint, groupid, blocking=True):
```
Comment (mrjerryjohns) [inline src/controller/python/chip/ChipDeviceCtrl.py:589]: Why do we need this? We should only be supporting the REPL approach going forward.

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 3]
```diff
@@ -530,6 +616,14 @@ def ZCLSubscribeAttribute(self, cluster, attribute, nodeid, endpoint, minInterva
             raise UnknownAttribute(cluster, attribute)
         return asyncio.run(self.ReadAttribute(nodeid, [(endpoint, req)], reportInterval=(minInterval, maxInterval)))
 
+    def ZCLSubscribeEvent(self, cluster, event, nodeid, endpoint, minInterval, maxInterval, blocking=True):
```
Comment (mrjerryjohns) [inline src/controller/python/chip/ChipDeviceCtrl.py:619]: Not needed (see above comment).

File: src/controller/python/chip/clusters/Attribute.py [hunk 4]
```diff
@@ -246,6 +372,50 @@ def handleAttributeData(self, path: AttributePath, status: int, data: bytes):
...
+    def _handleEventData(self, header: EventHeader, data: bytes):
+        try:
+            path = EventPath(
+                ClusterId=header.ClusterId, EventId=header.EventId)
+            eventType = _EventIndex.get(str(path), None)
+            eventValue = None
+            tlvData = chip.tlv.TLVReader(data).get().get("Any", {})
+            if eventType is None:
+                eventValue = ValueDecodeFailure(
+                    tlvData, LookupError("attribute schema not found"))
```
Comment (mrjerryjohns) [inline src/controller/python/chip/clusters/Attribute.py:384]:
```suggestion
                    tlvData, LookupError("event schema not found"))
```

File: src/controller/python/chip/clusters/Attribute.py [hunk 5]
```diff
@@ -246,6 +372,50 @@ def handleAttributeData(self, path: AttributePath, status: int, data: bytes):
...
+                    logging.error(ex)
+                    eventValue = ValueDecodeFailure(
+                        tlvData, ex)
+
+                    # If we're in debug mode, raise the exception so that we can better debug what's happening.
+                    if (builtins.enableDebugMode):
+                        raise
+
+            with self._resLock:
+                if self._res['Events'].has_key(path):
```
Comment (mrjerryjohns) [inline src/controller/python/chip/clusters/Attribute.py:402]: An event stream is exactly that - a stream of events. That means that even if there are duplicated events for the same path, each and every one of them should be visible.

This makes it behave like attributes that patch up the cache, which is not what we want.

File: src/controller/python/chip/clusters/Attribute.py [hunk 6]
```diff
@@ -65,17 +65,92 @@ def __hash__(self):
...
     ListIndex: int = None
 
 
+@dataclass
+class EventHeader:
+    EndpointId: int = None
+    ClusterId: int = None
+    EventId: int = None
+    EventNumber: int = None
+    Priority: int = None
```
Comment (mrjerryjohns) [inline src/controller/python/chip/clusters/Attribute.py:111]: Should be strongly typed as an enum.

File: src/controller/python/chip/clusters/Attribute.py [hunk 7]
```diff
@@ -65,17 +65,92 @@ def __hash__(self):
...
+    Timestamp: int = None
+    TimestampType: int = None
```
Comment (mrjerryjohns) [inline src/controller/python/chip/clusters/Attribute.py:113]: Enum.

**Review by mrjerryjohns — CHANGES_REQUESTED** [review]

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 2 above)
Comment (yunhanw-google) [reply src/controller/python/chip/ChipDeviceCtrl.py:589]: sure, will remove it.

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 3 above)
Comment (yunhanw-google) [reply src/controller/python/chip/ChipDeviceCtrl.py:619]: sure, will remove it

File: src/controller/python/chip/clusters/Attribute.py (hunk 5 above)
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py:402]: I am using the path as key, the value is a list which store a list of events with same path,
this logic here would firstly check if that event path has been stored, we would find that list and append event to that list.

File: src/controller/python/chip/clusters/Attribute.py (hunk 6 above)
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py]: Done, thanks

File: src/controller/python/chip/clusters/Attribute.py (hunk 7 above)
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py]: Done, thanks

File: src/controller/python/chip/clusters/Attribute.py (hunk 4 above)
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py]: done, thanks

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 3 above)
Comment (yunhanw-google) [reply src/controller/python/chip/ChipDeviceCtrl.py]: removed, thanks

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 2 above)
Comment (yunhanw-google) [reply src/controller/python/chip/ChipDeviceCtrl.py]: removed, thanks

File: src/controller/python/chip/clusters/Attribute.py (hunk 5 above)
Comment (mrjerryjohns) [reply src/controller/python/chip/clusters/Attribute.py:413]: I don't think we should structure it that way - events should just be in a flat list sorted by the order in which they were emitted in time.

Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py]: done, thanks

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 8]
```diff
@@ -484,7 +484,83 @@ async def ReadAttribute(self, nodeid: int, attributes: typing.List[typing.Union[
...
+        eventLoop = asyncio.get_running_loop()
+        future = eventLoop.create_future()
+
+        device = self.GetConnectedDeviceSync(nodeid)
+        eves = []
+        for v in events:
+            endpoint = None
+            cluster = None
+            event = None
+            if v == ('*') or v == ():
```
Comment (electrocucaracha) [inline src/controller/python/chip/ChipDeviceCtrl.py:534]: nit
```suggestion
            if v in [('*'), ()]:
```

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 9]
```diff
@@ -484,7 +484,83 @@ async def ReadAttribute(self, nodeid: int, attributes: typing.List[typing.Union[
...
+                    raise ValueError("Unsupported Event Path")
+            else:
+                # endpoint + (cluster) event / endpoint + cluster
+                endpoint = v[0]
+                if issubclass(v[1], ClusterObjects.Cluster):
+                    cluster = v[1]
+                elif issubclass(v[1], ClusterAttribute.ClusterEventDescriptor):
+                    event = v[1]
+                else:
+                    raise ValueError("Unsupported Attribute Path")
```
Comment (electrocucaracha) [inline src/controller/python/chip/ChipDeviceCtrl.py:555]: What about using kind of `switch` approach?
```suggestion
            print(type(v))
            if type(v) is tuple:
                # not a wildcard
                if not v in [('*'), ()]:
                    # endpoint + (cluster) event / endpoint + cluster
                    endpoint = v[0]
                    if issubclass(v[1], ClusterObjects.Cluster):
                        cluster = v[1]
                    elif issubclass(v[1], ClusterAttribute.ClusterEventDescriptor):
                        event = v[1]
                    else:
                        raise ValueError("Unsupported Attribute Path")
            elif type(v) is int:
                endpoint = v
            elif issubclass(v, ClusterObjects.Cluster):
                cluster = v
            elif issubclass(v, ClusterObjects.ClusterEventDescriptor):
                event = v
            else:
                 raise ValueError("Unsupported Event Path")
```

File: src/controller/python/chip/ChipDeviceCtrl.py [hunk 10]
```diff
@@ -484,7 +484,83 @@ async def ReadAttribute(self, nodeid: int, attributes: typing.List[typing.Union[
...
+            eves.append(ClusterAttribute.EventPath(
+                EndpointId=endpoint, Cluster=cluster, Event=event))
+        res = self._ChipStack.Call(
+            lambda: ClusterAttribute.ReadEvents(future, eventLoop, device, self, eves, ClusterAttribute.SubscriptionParameters(reportInterval[0], reportInterval[1]) if reportInterval else None))
+        if res != 0:
+            raise self._ChipStack.ErrorToException(res)
```
Comment (electrocucaracha) [inline src/controller/python/chip/ChipDeviceCtrl.py:561]: nit
```suggestion
        if self._ChipStack.Call(
            lambda: ClusterAttribute.ReadEvents(future, eventLoop, device, self, eves, ClusterAttribute.SubscriptionParameters(reportInterval[0], reportInterval[1]) if reportInterval else None)) != 0:
            raise self._ChipStack.ErrorToException(res)
```

File: src/controller/python/chip/clusters/Attribute.py [hunk 11]
```diff
@@ -65,17 +78,92 @@ def __hash__(self):
...
+        self.EndpointId = EndpointId
+        if Cluster is not None:
+            # Wildcard read for a specific cluster
+            if (Event is not None) or (ClusterId is not None) or (EventId is not None):
+                raise Warning(
+                    "Event, ClusterId and AttributeId is ignored when Cluster is specified")
+            self.EventId = Event.id
+            return
+        if Event is not None:
+            if (ClusterId is not None) or (EventId is not None):
```
Comment (electrocucaracha) [inline src/controller/python/chip/clusters/Attribute.py:97]: I just wondering if you rely on their [truthy](https://docs.python.org/3/library/stdtypes.html#truth-value-testing) values:
```suggestion
        if Cluster:
            # Wildcard read for a specific cluster
            if Event or ClusterId or EventId:
                raise Warning(
                    "Event, ClusterId and AttributeId is ignored when Cluster is specified")
            self.EventId = Event.id
            return
        if Event:
            if ClusterId or EventId :
```

File: src/controller/python/chip/clusters/Attribute.py [hunk 12]
```diff
@@ -65,17 +78,92 @@ def __hash__(self):
...
+        self.EndpointId = EndpointId
+        if Cluster is not None:
+            # Wildcard read for a specific cluster
+            if (Event is not None) or (ClusterId is not None) or (EventId is not None):
+                raise Warning(
+                    "EventId, ClusterId and EventId is ignored when Cluster is specified")
+            self.ClusterId = Cluster.id
+            return
+        if Event is not None:
+            if (ClusterId is not None) or (EventId is not None):
```
Comment (electrocucaracha) [inline src/controller/python/chip/clusters/Attribute.py:138]: ditto: line 89
```suggestion
        if Cluster:
            # Wildcard read for a specific cluster
            if Event or ClusterId or EventId:
                raise Warning(
                    "EventId, ClusterId and EventId is ignored when Cluster is specified")
            self.ClusterId = Cluster.id
            return
        if Event:
            if ClusterId or EventId:
```

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 10 above)
Comment (yunhanw-google) [reply src/controller/python/chip/ChipDeviceCtrl.py:561]: need save res result, and raise the exception with res, thanks

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 8 above)
Comment (yunhanw-google) [reply src/controller/python/chip/ChipDeviceCtrl.py:533]: done, thanks

File: src/controller/python/chip/ChipDeviceCtrl.py (hunk 9 above)
Comment (yunhanw-google) [reply src/controller/python/chip/ChipDeviceCtrl.py:554]: this logic is similar to what we use for attribute, and would update it in follow-up PR with attribute together.

File: src/controller/python/chip/clusters/Attribute.py (hunk 12 above)
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py:138]: EventId could be 0, so we would not this way, thanks for suggestion.

File: src/controller/python/chip/clusters/Attribute.py (hunk 11 above)
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py]: EventId could be 0, so we would not this way, thanks for suggestion.

File: src/controller/python/chip/clusters/Attribute.py [hunk 13]
```diff
@@ -65,17 +78,92 @@ def __hash__(self):
...
 @dataclass
 class AttributePathWithListIndex(AttributePath):
     ListIndex: int = None
 
 
+@dataclass
+class EventHeader:
+    EndpointId: int = None
+    ClusterId: int = None
+    EventId: int = None
```
Comment (mrjerryjohns) [inline src/controller/python/chip/clusters/Attribute.py:122]: We can make this a bit more usable for our users - Kind of like what I'm about to do for attributes, we should instead be using the event cluster object **type** itself as a replacement for expressing raw numbers like this. That would then replace event ID and cluster ID in one swoop.

File: src/controller/python/chip/clusters/Attribute.py [hunk 14]
```diff
@@ -65,17 +78,92 @@ def __hash__(self):
...
+    EventNumber: int = None
+    Priority: EventPriority = None
+    Timestamp: int = None
+    TimestampType: EventTimestampType = None
+
+    def __init__(self, EndpointId: int = None, Cluster=None, Event=None, ClusterId=None, EventId=None, EventNumber=None, Priority=None, Timestamp=None, TimestampType=None):
+        self.EndpointId = EndpointId
+        if Cluster is not None:
+            # Wildcard read for a specific cluster
```
Comment (mrjerryjohns) [inline src/controller/python/chip/clusters/Attribute.py:131]: Why is anything about wildcards even mentioned here? Isn't this supposed to be paths in event headers, which are always concrete?

File: src/controller/python/chip/clusters/Attribute.py [hunk 15]
```diff
@@ -65,17 +78,92 @@ def __hash__(self):
...
 @dataclass
 class AttributeStatus:
     Path: AttributePath
     Status: Union[chip.interaction_model.Status, int]
 
 
+@dataclass
+class EventStatus:
+    Header: EventHeader
+    Status: Union[chip.interaction_model.Status, int]
```
Comment (mrjerryjohns) [inline src/controller/python/chip/clusters/Attribute.py:164]: Comments please as to why this is a union.

File: src/controller/python/chip/clusters/Attribute.py (hunk 14 above)
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py]: thanks for catching up this issue, yes, removed and updated, thanks

File: src/controller/python/chip/clusters/Attribute.py (hunk 13 above)
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py:122]: agree, maybe we can follow-up this?

File: src/controller/python/chip/clusters/Attribute.py (hunk 15 above)
Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py:150]: I think it is fine to remove union. thanks

File: src/controller/python/chip/clusters/Attribute.py [hunk 16]
```diff
@@ -323,6 +472,15 @@ def _OnReadAttributeDataCallback(closure, endpoint: int, cluster: int, attribute
         EndpointId=endpoint, ClusterId=cluster, AttributeId=attribute), status, dataBytes[:])
 
 
+@_OnReadEventDataCallbackFunct
+def _OnReadEventDataCallback(closure, endpoint: int, cluster: int, event: int, number: int, priority: int, timestamp: int, timestampType: int, data, len):
+    dataBytes = ctypes.string_at(data, len)
+    path = EventPath(ClusterId=cluster, EventId=event)
+    eventType = _EventIndex.get(str(path), None)
```
Comment (mrjerryjohns) [inline src/controller/python/chip/clusters/Attribute.py:479]: This _can_ fail. We should do something similar to what we do for attributes:

```
            attributeType = _AttributeIndex.get(str(AttributePath(
                ClusterId=path.ClusterId, AttributeId=path.AttributeId)), None)
            attributeValue = None
            tlvData = chip.tlv.TLVReader(data).get().get("Any", {})
            if attributeType is None:
                attributeValue = ValueDecodeFailure(
                    tlvData, LookupError("attribute schema not found"))
```

That should happen in `handleEventData`. Consequently, you should:
1. Avoid initializing `Event` in the `EventHeader` built below.
2. Pass in the event ID into `handleEventData` to permit it to properly do the look up.

Comment (yunhanw-google) [reply src/controller/python/chip/clusters/Attribute.py]: thanks, just update, pass EventPath into handleEventData to permit it to properly do the look u

**Review by mrjerryjohns — APPROVED** [review]

**Comment by andy31415** [comment]
fast track: pr created and approved by domain owners,  has been up  for a long time, touches python (i.e. not functionality) with changes that  are non-python not behavioral

**Pull request merged**